packages/core/test/unit/writer.test.ts

```diff
@@ -4,36 +4,48 @@
 // License text available at https://opensource.org/licenses/MIT
 
 import {
+  ServerResponse,
   OperationRetval,
   writeResultToResponse,
 } from '../..';
-import {expect, ShotRequest, ShotResponse, ShotCallback, ShotResponseCtor} from '@loopback/testlab';
 
+import {
+  expect,
+  ShotRequest,
+  ShotResponse,
+  ShotObservedResponse,
+} from '@loopback/testlab';
 
 describe('writer', () => {
-  it('writes result to response', async () => {
-    const req = new ShotRequest({url: '/'});
-    const res = new ShotResponse(req, () => {});
-    // content-type should be undefined since it's not set in the response yet.
-    expect(res.getHeader('content-type')).to.eql(undefined);
-    await writeResultToResponse(res, 'Joe');
+  let response: ServerResponse;
+  let observedResponse: Promise<ShotObservedResponse>;
+
+  beforeEach(setupResponseMock);
+
+  it('writes string result to response as text', async () => {
+    writeResultToResponse(response, 'Joe');
+    const result = await observedResponse;
+
     // content-type should be 'application/json' since it's set into the response in writer.writeResultToResponse()
-    expect(res.getHeader('content-type')).to.eql('text/plain');
-
-    // TODO [rashmi] below code checks if response body has "Joe". However, res.on(..) is not getting
-    // invoked. Is this because writer.writeResultToResponse() already invokes response.end()?
-    // If this is the reason, we can't really check verify the content of the response body in the unit test?
```

Comment:
`ShotResponse`'s constructor takes a callback that's invoked once the response was fully written, and an object containing the response data is parsed to that callback.

```js
new ShotResponse(req, res => expect(res.payload).to.equal('Joe'));
```

I am adding a bit of helping code in this test to make that callback-based API easier to consume in `await` style, see below. We should probably move that helper to testlab at some point in the future.